File: javascript/src/day03.js
const fs = require('fs');

function paint(cloth, patch) {
  let isUnbroken = true;
  for(let y = patch.y; y < patch.y + patch.height; y++)
    for (let x = patch.x; x < patch.x + patch.width; x ++)
      if (cloth.fabric[y][x]) {
        cloth.unBrokenPatches.delete(cloth.fabric[y][x]);
        cloth.fabric[y][x] = -1;
        isUnbroken = false;
      } else {
        cloth.fabric[y][x] = patch.id;
      }

  if (isUnbroken)
    cloth.unBrokenPatches.add(patch.id);
}

function createCloth(width, height) {
  const createRow = () => Array(width).fill(0);
  const fabric = Array.from({ length: height }, createRow);
  return { fabric, unBrokenPatches: new Set() };
}

function getOverlappingCount(patches) {
  const cloth = createCloth(1000, 1000);
  patches.forEach(patch => paint(cloth, patch));
  let count = 0;
  for(let y = 0; y < 1000; y++)
    cloth.fabric[y].forEach(n => { if (n === -1) count++; });
  return count;
}

function readPatches() {
  const parsePatch = line => {
    const [, id, x, y, width, height] = line.split(/\D+/g).map(Number);
    return { id, x, y, width, height };
  };

  return fs.readFileSync('../input/day03.txt', 'utf-8')
    .split('\n')
    .map(parsePatch);
}

function part1() {
  const patches = readPatches();
  const count = getOverlappingCount(patches);
  console.log('There are', count, 'square inches that are covered by two or more patches');
}

function part2() {
  const patches = readPatches();
  const cloth = createCloth(1000, 1000);
  patches.forEach(patch => paint(cloth, patch));
  const [result] = cloth.unBrokenPatches;
  console.log(`The only unbroken patch is #${result}`);
}

module.exports = function solve() {
  console.log();
  console.log('Day 3');
  part1();
  part2();
}

File: haskell/src/Day03.hs
{-# LANGUAGE TupleSections #-}
module Day03 where

import Data.Array
import qualified Data.Ix as Ix
import Data.Set (Set)
import qualified Data.Set as Set

import Util

data Patch = Patch { _id:: Int
                   , _x:: Int
                   , _y:: Int
                   , _width:: Int
                   , _height:: Int
                   } deriving (Show, Eq)

type Fabric = Array (Int, Int) (Set Int)
type Cloth = Fabric

parseInput :: String -> [Patch]
parseInput input =
  let
    createPatch (pid : x : y : w : h : _) = Patch pid x y w h
  in
    map (createPatch . parseInts) (lines input)

createCloth :: Cloth
createCloth =
  let
    fabric = listArray ((0,0), (1000, 1000)) (repeat Set.empty)
  in
   fabric

paintPatch :: Cloth -> Patch -> Cloth
paintPatch fabric (Patch pid x y w h) =
  let
    paintCell :: Set Int -> Int -> Set Int
    paintCell = flip Set.insert
    patchCells :: [((Int, Int), Int)]
    patchCells = map (, pid) $ Ix.range ((x, y), (x + w - 1, y + h - 1))
  in
    accum paintCell fabric patchCells

countMultiPatchCells :: Cloth -> Int
countMultiPatchCells fabric =
  let
    cells = elems fabric
    multiPatchCells = filter ((> 1) . length) cells
  in
    length multiPatchCells

part1 :: IO ()
part1 = do
  input <- readFile "../input/day03.txt"
  let patches = parseInput input
  let cloth = foldl paintPatch createCloth patches
  let multis = countMultiPatchCells cloth
  putStrLn ("There are " ++ show multis ++ " cells with more than one patch")

-- part1 :: IO ()
-- part1 = do
--   cloth <- createCloth
--   let patch = Patch 15 2 3 4 5
--   paintPatch cloth patch
--   print 5

solve :: IO ()
solve = do
  putStrLn ""
  putStrLn "Day 3"
  part1
